File: bin/gvn_update_mapping_db.py
#!/usr/bin/env python3

# Description: GVN - Git sVN - git configuration and more for easier usage of git-svn, test suite
# Author:      jlechnar
# Licence:     GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007
# Source:      https://github.com/jlechnar/gvn

import argparse
import re

import os
import sys

from inspect import getsourcefile
from os.path import realpath
from os.path import dirname

# add path where script is located to be able to read all submodules
# submodules are expected to be located / must be located parallel to this file !
sys.path.insert(0, dirname(realpath(getsourcefile(lambda:0))))

from tools_c import *
from gvn_exceptions import *
from gvn_colors import *
from gvn_base import *

# -----------------------------------------------------------
if __name__ == '__main__':
            
    try:
        # -------------------------------------
        verbose = 0
        tools = tools_c() 
        
        # -------------------------------------
        parser = argparse.ArgumentParser()
        parser.add_argument("--debug", "-d", help="enable debug", action="store_true")
        args = parser.parse_args()

        # ---------------------
        base = gvn_base(tools, args.debug, verbose)

        base.init_hash_to_svn_rev()

        command="git get-dot-git-path"
        result = subprocess.run(command.split(), stdout=subprocess.PIPE)
        db_filename = result.stdout.decode().rstrip()
        base.write_to_json_file(db_filename)
        
        # ---------------------
    except KeyboardInterrupt:
        print("\n")
        tools.error("User Abort detected.")
        #    except ToolException as e:
        #        the_tools.error("Uncatched Tool Error detected:\n " + str(e)) # this should not happen but catch it here just in case
    except GVNException as e:
        tools.error("Error detected:\n " + str(e))
        #    except BaseException as e:
        #        the_tools.error("Fatal Uncatched Error detected:\n " + str(e))

File: bin/gvn_update_mapping_db.pl
#!/usr/bin/perl -w

# Description: GVN - Git sVN - git configuration and more for easier usage of git-svn, test suite
# Author:      jlechnar
# Licence:     GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007
# Source:      https://github.com/jlechnar/gvn

use strict;
use warnings;

use Getopt::Std;

use List::Util qw( min max );

use Term::ANSIColor;

my %opts;
getopts("d", \%opts) or die;

# ------------------------------
my $hash = "";
my $revision = "";
my $source = "";
my $revision_length = 0;

my %db = ();

open(my $fh, '-|', 'git log --all --no-color') or die $!;
while (my $line = <$fh>) {
    chomp $line;

    if($line =~ /^\s*commit\s+(\S+)(\s+|$)/) {
      $hash = $1;
    } elsif($line =~ /^\s+git-svn-id:\s+([^\@]+)\@(\d+)\s+/) {
      $source = $1;
      $revision = $2;
      if($hash ne "") {
        $db{$hash} = $revision;
	$revision_length = max(length($revision), $revision_length);
        if($opts{d}) {
          warn ("$hash -> $revision\n");
	}
      } else {
	die("FATAL: undefined hash\n");
      }
    }
}

my $dot_git_path=`git get-dot-git-path`;
chomp $dot_git_path;
my $db_filename="$dot_git_path/gvn/rev/db.list";

open(FILE, ">$db_filename") or die "Couldn't open file $db_filename for writing, $!";

print FILE "$revision_length\n";
for my $hash (keys %db) {
  my $revision = $db{$hash};
  print FILE "$hash $revision\n";
}

close(FILE);

